resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,127 @@
+  hadoopUGI: HadoopUGIUtil) extends HadoopConfigurationStep with Logging {
+    private var originalCredentials: Credentials = _
+    private var dfs : FileSystem = _
+    private var renewer: String = _
+    private var credentials: Credentials = _
+    private var tokens: Iterable[Token[_ <: TokenIdentifier]] = _
+
+    override def configureContainers(hadoopConfigSpec: HadoopConfigSpec): HadoopConfigSpec = {
+      val hadoopConf = SparkHadoopUtil.get.newConfiguration(submissionSparkConf)
+      if (!hadoopUGI.isSecurityEnabled) logDebug("Hadoop not configured with Kerberos")
+      val maybeJobUserUGI =
+        for {
+          principal <- maybePrincipal
+          keytab <- maybeKeytab
+        } yield {
+          // Not necessary with [Spark-16742]
+          // Reliant on [Spark-20328] for changing to YARN principal
+          submissionSparkConf.set("spark.yarn.principal", principal)
+          submissionSparkConf.set("spark.yarn.keytab", keytab.toURI.toString)
+          logDebug("Logged into KDC with keytab using Job User UGI")
+          hadoopUGI.loginUserFromKeytabAndReturnUGI(
+            principal,
+            keytab.toURI.toString)
+        }
+      // In the case that keytab is not specified we will read from Local Ticket Cache
+      val jobUserUGI = maybeJobUserUGI.getOrElse(hadoopUGI.getCurrentUser)
+      // It is necessary to run as jobUserUGI because logged in user != Current User
+      jobUserUGI.doAs(new PrivilegedExceptionAction[Void] {
+        override def run(): Void = {
+          originalCredentials = jobUserUGI.getCredentials
+          // TODO: This is not necessary with [Spark-20328] since we would be using
+          // Spark core providers to handle delegation token renewal
+          renewer = maybeRenewerPrincipal.getOrElse(jobUserUGI.getShortUserName)
+          credentials = new Credentials(originalCredentials)
+          hadoopUGI.dfsAddDelegationToken(hadoopConf, renewer, credentials)
+          tokens = credentials.getAllTokens.asScala
+          null
+        }})
+      if (tokens.isEmpty) logDebug("Did not obtain any Delegation Tokens")
+      val data = hadoopUGI.serialize(credentials)
+      val renewalInterval =
+        hadoopUGI.getTokenRenewalInterval(tokens, hadoopConf).getOrElse(Long.MaxValue)
+      val currentTime: Long = hadoopUGI.getCurrentTime
+      val initialTokenDataKeyName = s"$KERBEROS_SECRET_LABEL_PREFIX-$currentTime-$renewalInterval"
+      val uniqueSecretName =
+        s"$kubernetesResourceNamePrefix-$HADOOP_KERBEROS_SECRET_NAME.$currentTime"
+      val secretDT =
+        new SecretBuilder()
+          .withNewMetadata()
+            .withName(uniqueSecretName)
+            .withLabels(Map(KERBEROS_REFRESH_LABEL_KEY -> KERBEROS_REFRESH_LABEL_VALUE).asJava)
+            .endMetadata()
+            .addToData(initialTokenDataKeyName, Base64.encodeBase64String(data))
+        .build()
+      val bootstrapKerberos = new KerberosTokenConfBootstrapImpl(
```

Comment:
The `KerberosTokenConfBootstrapImpl` instance should be injected into this class. The unit test for this class should use a mock for that object.

Reply:
Good insight. Will do

Reply:
^ this is trickier as the `KerberosTokenConfBootstrapImpl` takes in the JobUserUGI which is determined in this class.

Reply:
You can have a constructor chain to address this:

1. Move the logic that determines the JobUserUGI to a utility object method.
2. Add a constructor that calls the JobUserUGI utility method.
3. Modify the class constructor to take KerberosTokenConfBootstrapImpl as a parameter. 
4. Let the first constructor from (2) call the class constructor with a KerberosTokenConfBootstrapImpl created using the JobUserUGI utility method output.